Refactor SCSS with variables

puntope opened the pull request:
### Changes proposed in this Pull Request:

Closes part of #1429

This PR implements the next changes:

- Create custom SCSS variables prefixed with `$gla-` for font size, dimensions and line-height.  
- Replace repetitive hardcoded props related to colors, sizes and fonts with WP variables or GLA variables.
- Remove redundant `_gla_variables.scss` file in favor of `_variables.scss`

### Detailed test instructions:

1. Check the changed files and see that the new variables are equivalent to the previous hardcoded values

### Additional details:

We added the next variables:

```
// Typography

$gla-font-smallest: 11px;
$gla-font-small: 14px;
$gla-font-small-medium: 20px;
$gla-font-medium: 24px;
$gla-font-big: 32px;

// Line Heights

$gla-line-height-smallest: 13px;
$gla-line-height-smaller: 16px;
$gla-line-height-small: 18px;
$gla-line-height-medium: 20px;
$gla-line-height-medium-large: 28px;
$gla-line-height-large: 32px;

// Dimensions
$gla-size-control-height: 36px;
$gla-size-small: 250px;
$gla-size-medium: 300px;

// Grid
$gla-grid-unit: 6px;
$gla-grid-unit-05: 0.5 * $gla-grid-unit;	// 3px
```

We replaced some other values with WP variables located in

`node_modules/@wordpress/base-styles/_variables.scss`

Especially, using the `grid-units` as well as `$icon-size`, borders and dimensions 

### Changelog entry

> Tweak - Refactor SCSS variables

puntope committed:
Update file name

puntope committed:
Refactor font sizes and line height values into variables

puntope committed:
Refactor font sizes and line height values into variables

puntope committed:
Refactor font sizes and line height values into variables

puntope committed:
Rename _variables scss files

puntope committed:
Merge branch 'refactor/scss-typography' into refactor/scss-variables

# Conflicts:
#	js/src/components/contact-information/phone-number-card/phone-number-card.scss
#	js/src/components/supported-country-select/supported-country-select.scss
#	js/src/product-feed/product-feed-table-card/index.scss
#	js/src/wcdl/radio-helper-text/index.scss

puntope committed:
Merge branch 'refactor/scss-variable-file' into refactor/scss-variables

# Conflicts:
#	js/src/css/abstracts/_variables.scss

ianlin commented on js/src/css/abstracts/_variables.scss:
```diff
@@ -2,3 +2,29 @@
+$gla-font-small: 14px;
+$gla-font-small-medium: 20px;
+$gla-font-medium: 24px;
+$gla-font-big: 32px;
```
💅 Maybe rename this to `$gla-font-large` to align with the naming convention in this file?

puntope replied:
addressed here https://github.com/woocommerce/google-listings-and-ads/pull/1578/commits/980d8f432d105be3def1a0c848822d10eedef34e

eason9487 commented on js/src/components/help-popover/index.scss:
```diff
@@ -21,14 +21,14 @@ $button-color: #949494;
 
 		@include break-large {
-			width: 360px;
+			width: $modal-min-width;
```
❓ Although the size is the same, using a variable named `$modal-min-width` here for the width of popover content looks fragile.

puntope replied:
Can you elaborate what do you mean with fragile?

eason9487 replied:
Considering it's a popover component, using a global variable that is declared in an external source and named for another type of component makes it look fragile. If its value changes someday, the purpose of the change from external sources is based on its applicability to their modal component, and there is uncertainty as to whether its new value still applies here. Moreover, GLA would not be aware of the change when it occurs.

puntope replied:
I agree that "modal" suffix is not optimal in terms of lack of flexibility. Kind of the same as discussed here https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920847312

> If its value changes someday, the purpose of the change from external sources is based on its applicability to their modal component, and there is uncertainty as to whether its new value still applies here. Moreover, GLA would not be aware of the change when it occurs.

I'm not sure about that fragility is an issue since that applies for all the SCSS variables (colors, gaps, spacing, fonts...)  as well as JS components regarding Wordpress and Woo. 

Another "this is not an issue" point is that I see them being used widely over in several places. 

<img width="1326" alt="Screenshot 2022-07-15 at 12 20 36" src="https://user-images.githubusercontent.com/5908855/179183562-af816a63-b117-4151-87a6-f58bd3fc95f1.png">

eason9487 replied:
> I'm not sure about that fragility is an issue since that applies for all the SCSS variables (colors, gaps, spacing, fonts...) as well as JS components regarding Wordpress and Woo.

It depends on the association between the names of the external variables and the GLA styles being applied. Popover and modal components are not the same from the perspective of layout, so I thought their association is relatively low.

> Another "this is not an issue" point is that I see them being used widely over in several places.

Sorry, I'm not pretty sure it's a reasonable point. All `$break-*` variables are breakpoints, using them with `@media` syntax is reasonable. But I don't see the values of breakpoints having an association with the dimensional width of a component's layout. Even though breaking points are almost never changed, the semantics are confusing.

eason9487 commented on js/src/components/help-popover/index.scss:
```diff
@@ -21,14 +21,14 @@ $button-color: #949494;
 
 		@include break-xlarge {
-			width: 480px;
+			width: $break-mobile;
```
❓ Same as above.

ianlin reviewed (approved):
Thanks for the big change! I've verified the hardcoded values are replaced by the scss variables, and I don't see the error like `SassError: Undefined variable.` when building the project. LGTM 👍

eason9487 commented on js/src/product-feed/submission-success-guide/index.scss:
```diff
@@ -66,12 +66,12 @@
+				padding: 0 $grid-unit-20;
 				box-sizing: border-box;
-				border-radius: 3px;
+				border-radius: $gla-grid-unit-05;
```
❔ I think it's ok to keep `border-radius` using a hardcoded value here, or maybe create another variable for it since it's a bit confusing to associate a *grid unit* to `border-radius`.

puntope replied:
Whats the issue using grid with border rad?

eason9487 replied:
`border-radius` is a style about the border but not the layout. Just like using a \<textarea\> to display a text block, it works but is confusing.

puntope replied:
Ok get it. So I created a new variable `$gla-border-radius` in here just for those. 

https://github.com/woocommerce/google-listings-and-ads/pull/1578/commits/694e94646af441a03955c6718a481e3427327dce

eason9487 commented on js/src/components/account-card/index.scss:
```diff
@@ -45,8 +45,8 @@
-			min-height: 36px;
-			padding: 6px 12px;
+			min-height: $gla-size-control-height;
+			padding: $gla-grid-unit $grid-unit-15;
```
🧵 
The variables for grid units declared in Gutenberg should be useful enough. I think GLA doesn't need to declare another list of grid units itself. For the 6px value, it could be `$grid-unit-15 / 2`.

puntope replied:
Addressed in last commits

eason9487 commented on js/src/components/contact-information/phone-number-card/verification-code-control.scss:
```diff
@@ -1,7 +1,7 @@
 	// Verification code input
 	.components-input-control .components-input-control__container .components-input-control__input {
-		width: 40px;
+		width: $grid-unit-50;
```
❔ Looks like the grid unit is more intently to be used in spacing. Would here be better to use a dimension value?

puntope replied:
I see all over the codebase of WP / WC using grid units for sizes. I see them as a grid relative unit size, so makes sense for me to use them also for width, height etc...

eason9487 commented on js/src/components/free-listings/configure-product-listings/shipping-time/shipping-time-setup/index.scss:
```diff
@@ -5,7 +5,7 @@
 
 		.countries-time-input-form {
-			max-width: 300px;
+			max-width: $gla-size-medium;
```
💅 
`$gla-size-small` and `$gla-size-medium` are general naming but most dimension usages are not general layouts. Perhaps we could use a smaller and more specific name.

puntope replied:
We can use maybe` $gla-layout-size-medium` ` $gla-layout-size-small` ? Suggestions welcome

eason9487 replied:
Probably I would use `$gla-width-shipping-form`.

puntope replied:
Problem with that is in case we want to use the size for another form, a modal or any other size definition it won't fit. 

So I would suggest to go for a more generic and flexible naming.

eason9487 replied:
> Problem with that is in case we want to use the size for another form, a modal or any other size definition it won't fit.

Yes, I meant they (modal, input, popover) could have their own dimensional variable rather than sharing the same one. And it's okay to keep it as is.

eason9487 commented on js/src/components/stepper/top-bar/index.scss:
```diff
@@ -1,9 +1,9 @@
-	min-height: 64px;
-	background-color: #fff;
-	box-shadow: 0 -1px 0 0 #ccc inset;
+	min-height: $grid-unit-40 * 2;
```
❔ Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920843375

eason9487 commented on js/src/components/app-modal/index.scss:
```diff
@@ -1,6 +1,6 @@
 .app-modal {
 	@include break-large {
-		width: 600px;
+		width: $break-small;
```
❓ Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920819280.

eason9487 commented on js/src/dashboard/all-programs-table-card/edit-program-button/edit-program-prompt-modal/index.scss:
```diff
@@ -1,5 +1,5 @@
 .gla-edit-program-prompt-modal {
 	@include break-small {
-		max-width: 360px;
+		max-width: $modal-min-width;
```
❔ Using modal-**min**-width variable for the **max**-with property looks a bit confusing.

eason9487 commented on js/src/get-started-page/get-started-card/index.scss:
```diff
@@ -22,11 +22,11 @@
 		flex-direction: column;
 		align-items: flex-start;
-		gap: 20px;
+		gap: calc(#{$grid-unit-50} / 2);
```
💅 A primitive sass variable can be calcauled directly.
```suggestion
		gap: $grid-unit-50 / 2;
```

puntope replied:
This creates a warning when running `npm run start`

eason9487 replied:
Got it. Looks like it could still be simplified a little bit with `gap: calc($grid-unit-50 / 2);`.

puntope replied:
<img width="1279" alt="Screenshot 2022-07-14 at 16 44 20" src="https://user-images.githubusercontent.com/5908855/178985486-269afa1a-bc06-401d-9cd3-d7e7349b7e5c.png">

eason9487 replied:
Not sure why but I didn't get that warning.

#### Change and rebuild

https://user-images.githubusercontent.com/17420811/179137245-573654a7-ef8e-4a5c-8a23-063ce4c59c9a.mp4

#### `npm start`

![image](https://user-images.githubusercontent.com/17420811/179137525-1e311d83-b148-468e-8ec6-d8e645ef20d7.png)

puntope replied:
> Not sure why but I didn't get that warning.

The warning is when doing `gap: $grid-unit / 2` if there is inside a calc it works. Seems like `/` is also being used as a separator.

More info:  https://sass-lang.com/documentation/breaking-changes/slash-div

Btw. I added the change removing `#{}` in this commit https://github.com/woocommerce/google-listings-and-ads/pull/1578/commits/1ed4470056ffcf3759ac1993670f5a7e85ed01dc

eason9487 replied:
Thanks for the adjustments.

> The warning is when doing `gap: $grid-unit / 2` if there is inside a calc it works. Seems like `/` is also being used as a separator.

I thought https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r921112923 was replying to  https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r921037788.

eason9487 commented on js/src/product-feed/submission-success-guide/index.scss:
```diff
@@ -38,9 +38,9 @@
-				min-width: 20px;
-				height: 20px;
-				padding: 6px;
+				min-width: calc(#{$grid-unit-50} / 2);
+				height: calc(#{$grid-unit-50} / 2);
```
💅 
```suggestion
				min-width: $grid-unit-50 / 2;
				height: $grid-unit-50 / 2;
```

puntope replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r921025753

eason9487 commented on js/src/setup-ads/ads-stepper/index.scss:
```diff
@@ -2,15 +2,15 @@
-		height: 64px;
-		background-color: #fff;
-		box-shadow: inset 0 -1px 0 #ccc;
+		height: $grid-unit-40 * 2;
```
❔ Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920843375.

eason9487 commented on js/src/setup-mc/setup-stepper/index.scss:
```diff
@@ -2,15 +2,15 @@
-		height: 64px;
-		background-color: #fff;
-		box-shadow: inset 0 -1px 0 #ccc;
+		height: $grid-unit-40 * 2;
```
❔ Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920843375.

eason9487 commented on js/src/wcdl/select-control/index.scss:
```diff
@@ -54,7 +54,7 @@
 				.woocommerce-tag__text {
 					padding: 0 $grid-unit-10;
-					border-radius: 12px;
+					border-radius: $grid-unit-15;
```
❔ Same as https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920833917.

puntope committed:
Replace name of gla-font-large

puntope committed:
Remove gla-grid variables and use calcs instead

puntope committed:
Fix format css

puntope committed:
Remove #{} for referencing SCSS variables inside calc

puntope committed:
Adding gla_border_radius variable

puntope commented:
Howdy again @eason9487 I added a few commits addressing some of your ❓ / 💅  and added more discussion on others. Waiting for your feedback. thanks

eason9487 reviewed (commented):
Thanks for addressing my concerns. The only thing I am still not pretty sure about whether the uses are a good idea is applying breakpoint variables to `width`:

- https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920819723
- https://github.com/woocommerce/google-listings-and-ads/pull/1578#discussion_r920866284

puntope committed:
Refactor width based variables

puntope commented:
> The only thing I am still not pretty sure about whether the uses are a good idea is applying breakpoint variables to `width`:

I addressed your concerns here https://github.com/woocommerce/google-listings-and-ads/pull/1578/commits/57d93214c06e1acdb600de7516896ccba8320f6d

In brief I rename/created a set of variables for width dimensions.

eason9487 reviewed (approved):
Thanks for the further adjustments.
